File: pages/refactorings/replace-mutable-fields-with-lenses/01-rename-var/Before.scala
// Uppercase the variable of one Var node: a hand-written match
// that rebuilds the hit branch and lets every other shape pass.
object Before:
  case class Var(name: String, ref: Int)
  enum Expr:
    case EVar(v: Var)
    case EApp(f: Expr, x: Expr)
    case ELam(bind: String, body: Expr)

  import Expr.*

  def upperVarName(e: Expr): Expr = e match
    case EVar(v) => EVar(v.copy(name = v.name.toUpperCase))
    case other   => other

File: pages/refactorings/replace-mutable-fields-with-lenses/01-rename-var/Before.hs
-- Uppercase the variable of one Var node: a hand-written match
-- that rebuilds the hit branch and lets every other shape pass.
module Before where

import Data.Char (toUpper)

data Var   = Var   { vName :: String, vRef :: Int }
  deriving (Eq, Show)
data Expr  = EVar Var | EApp Expr Expr | ELam String Expr
  deriving (Eq, Show)

upperVarName :: Expr -> Expr
upperVarName (EVar v) = EVar v { vName = map toUpper (vName v) }
upperVarName e        = e
